### ISSUE TITLE
Extend get_entities interface with result_types and streaming VS async resonses

### ISSUE THREAD

**Issue by moolitayer** [abonas/kubeclient#263]
Since we have a few[1][2] efforts going in the same direction that have implication on our external API
@cben @ilackarms and myself brainstormed to come up with
- minimal interface to allow most options
- implementation for the actual get_entities methods to accommodate most options
- what is the minimum we have to do now, what can we do later

### What do we want to allow?
Currently `get_entities` is synchronous and returns ROS wrapping JSON objects for a complete response after a request response cycle complete. These are all viable options:
1 synchronous get whole raw response ([2])
2 get each raw chunk when available
3 synchronous get whole JSON response
4 get each JSON object when available
5 synchronous get whole ROS response (current default)
6 get each ROS object when available ([1])

### How should our new interface look like?
1 `get_pods(:result_type[3] => :raw)`
2 `get_pods(:result_type => :raw) do |chunk|
...
end
`
3 `get_pods(:result_type => :json)`
4 `get_pods(:result_type => :json) do |json_pod|
...
end
`
5 `get_pods()` # :result_type => ros is our current default
6 `get_pods() do |ROS_pod|
...
end` # :result_type => ros is our current default

### What is the minimum we can do now?
- [1] internal get_entities always serves raw chunks since that is the lowest denominator, need to also accommodate current behavior - option 5 and option 6 if block is given
- [2] implement 1 and 2
- touch only get, not watch 

cc @ilackarms @cben 
@simon3z @abonas @NickLaMuro @grosser if you have no objection we like to proceed this way ASAP

[1] https://github.com/abonas/kubeclient/pull/254
[2] https://github.com/abonas/kubeclient/pull/262
[3] the name `:result_type` is ephemeral, might change in Code review

**Comment by cben** [comment]
> internal get_entities always serves raw chunks since that is the lowest
denominator

Are `get_entity/ies`, `update_entity` considered private or part of public
API?
They're not in README, but also aren't `private` methods.
I actually thought of them as the "core API" and `get_pods` etc as
convenience additions.

**Comment by moolitayer** [comment]
Interesting @cben I never thought about it that way but it is valid - I guess for sure we need to maintain api in both the dynamically created get_pods and static get_entities. I don't remember seeing much usage of get_entities in example code snippets we got from users though.

**Comment by grosser** [comment]
sounds good to me ... result_type sounds a bit clunky to me ... maybe `as:` or `format:` would read nicer ... but I guess they could be confused with other things ...

**Comment by NickLaMuro** [comment]
> * [1] internal get_entities always serves raw chunks since that is the lowest denominator...

I guess I am less in favor of this way of thinking.  I personally think we shouldn't be stream parsing when we aren't taking advantage of it, because otherwise you are just introducing a performance bottleneck on the client with every form (1 through 6) above.  While it makes for simpler code to implement, it means that we are always bottlenecked by the stream parser (even if it is noop-ing).

I guess I would have to see some benchmarks before and after to see if implementing a `kubeclient` this way would be a performance detriment at scale.  My guess it would make it slower for the non-block forms.

> * [2] implement 1 and 2

I assume by this, you mean implement https://github.com/abonas/kubeclient/pull/254 and then https://github.com/abonas/kubeclient/pull/262, correct?  For what it is worth, as it stands right now, those two are mutually exclusive PRs, not only from a merge conflict perspective, but from a implementation standpoint.

https://github.com/abonas/kubeclient/pull/254, as part of it's implementation, assumes we will always be parsing the valid JSON returned from kubernetes on a 200 request, so there is no way that https://github.com/abonas/kubeclient/pull/262 has any benefit once that is in place.  This means that significant changes are required on top of 
https://github.com/abonas/kubeclient/pull/254 to make https://github.com/abonas/kubeclient/pull/262 possible again for them both to work in tandem.

Also, I see https://github.com/abonas/kubeclient/pull/254 as requiring a major version bump on `kubeclient`.  It is introducing an entirely new dependency into the gem, and while a new interface to all of these methods is probably fine for a minor release, adding a new dependency, in my opinion, is not.

* * *

What I intended with my changes in https://github.com/abonas/kubeclient/pull/262, while not communicated in the description, was to be a stopgap while something like you described in this proposal could be implemented in full in the next release.  It allows for get around the bulky-ness of Recursive-OpenStruct when that is not a viable option, only adds to the codebase in a minimal fashion to do so, and all existing defaults remain.

I think with this proposal, you are really suggesting a major re-write, and thus, a new major version of the gem.  This is fine, but probably will be a bit of time to implement, and shouldn't be something we consider safe users using something like `gem "kubeclient", "~> 2.4"` to just receive as part of a `bundle update`.

**Comment by moolitayer** [comment]
> I guess I am less in favor of this way of thinking. I personally think we shouldn't be stream parsing when we aren't taking advantage of it, because otherwise you are just introducing a performance bottleneck on the client with every form (1 through 6) above. While it makes for simpler code to implement, it means that we are always bottlenecked by the stream parser (even if it is noop-ing).

Good point. @ilackarms what do you think? I don't know what the numbers are it might be a good idea to measure. Sounds safer to only loop if we need to.

> What I intended with my changes in #262, while not communicated in the description, was to be a stopgap while something like you described in this proposal could be implemented in full in the next release. It allows for get around the bulky-ness of Recursive-OpenStruct when that is not a viable option, only adds to the codebase in a minimal fashion to do so, and all existing defaults remain.

Sorry for slowing you down @NickLaMuro, my main concern is that the API is consistent and minimal. 
I fear if we add :raw => true now, in a few month we will have to consider :ros => true, :json => true or have to change interface and support backward compatibility for a deprecated option. if #262 addresses that then it's fine by me

**Comment by moolitayer** [comment]
> sounds good to me ... result_type sounds a bit clunky to me ... maybe as: or format: would read nicer ... but I guess they could be confused with other things ...

Very clunky indeed ¯\_(ツ)_/¯. From your suggestions I liked `as`. It's more then output as it also has to do with the way we get the response (initially I suggested `output` like in kubectl but @ilackarms objected on these grounds if i'm not mistaken)

**Comment by NickLaMuro** [comment]
> Sorry for slowing you down @NickLaMuro, my main concern is that the API is consistent and minimal.

@moolitayer Oh, I am definitely not blocked by any of this.  There is actually a pretty easy way for me to get around not having this place if necessary (I will link to it in another section).


> I fear if we add :raw => true now, in a few month we will have to consider :ros => true, :json => true or have to change interface and support backward compatibility for a deprecated option

As I mentioned, https://github.com/abonas/kubeclient/pull/262 was simply a stop gap solution, and I fully expect it to be completely re-worked to support what you have suggested in this proposal (to be fair, I think you would have to do the save for #254 as well).

If you are concerned about the addition of `:raw => true` and part in #262 being changed, I would fully understand not wanting to introduce it "as is" (or even with changes) if it doesn't support the new format that you are considering here.  That said, as I mentioned perviously, I see the changes proposed here being something that would require a major version bump, and wouldn't be something you would release without going to `v3.0.0`.


> @NickLaMuro can you elaborate a little about how you plan to use this?

You actually [asked this](https://github.com/abonas/kubeclient/pull/262#issuecomment-324300388) in https://github.com/abonas/kubeclient/pull/262, but I would rather respond in one spot, since it it relevant to this discussion.

I had done some work a few months ago on testing out some json stream parsers (different from what was used in #254 ) to see if it was possible to extract data using kubeclient without ballooning in memory when downloading a large `/images` payload.  The data in there is probably a little skewed, because VCR is also taking up some memory there.  That said, just loaded the JSON output of the json returned from the cassette in a test script, and got the following results:

```console
$ cat testscript.rb
hashes = JSON.parse(File.open("openshift_images.json").read)
puts "Mem: #{((1024 * BigDecimal.new(`ps -o rss= -p #{Process.pid}`))/BigDecimal.new(1_048_576)).to_f}MiB"
puts hashes["items"].count
$ ruby -rbigdecimal -rjson testscript.rb
Mem: 677.296875MiB
9166
```

So that is probably a bit more accurate of what the memory would be at a minimum, possibly a bit more once they get converted to recursive openstruct objects.  Point being, the two "optimization" PRs are essentially taking a monkey patched version of what I did in #262, and downloading the raw data in a separate process (just to avoid any memory issues ruby might have), and then parsing them using a stream processor to extract the necessary data for each record.

Now, this was geared to be used for the openshift refresh in ManageIQ, which most people in this convo at the time of writing (besides Mr. Grosser) are working on.  For the refresh, if you are unaware, the parsing of the data from the provider (OpenShift) is actually not the slowest of the refresh, but committing to the DB is.

For something like #254 to have any affect on memory, we would have to keep the connection to Openshift open while stream parsing the data during the entire duration of committing the images data to the database, which in most case takes probably 50x or more longer to do than just downloading and parsing the data we get.

To further emphasis that point, the data that we have seen from larger instances where we are looking at 20k images can take OpenShift (not sure how much if this is the Kubernetes code base at this point) about 2+ minutes to process, and the json payload for that endpoint is over 800Mb in size, and about 95% of the total payload we receive from openshift over all requests (in this specific instance).

So #262 was simply a way of allowing us to download the endpoint in full, save it off to a file somewhere so it doesn't have to be in memory in full, and then implement a streaming solution outside of kubeclient that would then allow the application to parse the json at it's leisure using it's own stream parser, and still gain the memory benefits from that.

* * *

Sorry... this has become another :book:... so I will end this here.  Hopefully this answered your questions, but I can elaborate (more than I already have) if needed.

**Comment by moolitayer** [comment]
Thanks @NickLaMuro, read it yesterday but couldn't response back then

> To further emphasis that point, the data that we have seen from larger instances where we are looking at 20k images can take OpenShift (not sure how much if this is the Kubernetes code base at this point) about 2+ minutes to process, and the json payload for that endpoint is over 800Mb in size, and about 95% of the total payload we receive from openshift over all requests (in this specific instance).

I'm sure you know this and it's outside the point here but pruning openshift images can help with that.

Anyway I'm good to proceed with #262 with out any dependency, we can continue the discussion from there.

**Comment by NickLaMuro** [comment]
> I'm sure you know this and it's outside the point here but pruning openshift images can help with that.

Yup, for sure, though I don't see 20K images being something that could possibly exist in a large environment, even when pruning is happening.  But yes, there are other ways around the problem.

**Comment by cben** [comment]
> For something like #254 to have any affect on memory, we would have to
keep the connection to Openshift open while stream parsing the data during
the entire duration of committing the images data to the database, which in
most case takes probably 50x or more longer to do than just downloading and
parsing the data we get.

This is getting off-topic for kubeclient, but some scenarios (refresh
parser process all images, then drop most) would only need connection for
duration of refresh parser, which is just few minutes.  Anyway save file is
cool idea!  "Low-tech" but effective :-)
Which reminds me, rest-client 2.1 (rc) will have some API extensions for
streaming, including to a file.  Perhaps worth looking for API inspiration:
https://github.com/rest-client/rest-client#streaming-responses

**Comment by cben** [comment]
Thanks!  That looks good enough.
If we do want to something prettier, I'm sure we could contribute to
json-streamer rather than inventing a new wheel.

Not sure JSONpath would be ideal.  It's both too complex (a mini language
to parse) and not the most flexible.
I was thinking of ruby callback that gets current path - array of strings /
indexes.  And the aggregator is very close to that!

**Comment by cben** [comment]
An update to contradict [previous comments](https://github.com/abonas/kubeclient/issues/263#issuecomment-324013580):
We've decided with @moolitayer that `get_entities`, `update_entity` etc. methods are PRIVATE as long as they're UNDOCUMENTED.  E.g. we're going to change them in #292 and don't even think we have to mention that.
Everything undocumented is considered internal API, doesn't matter if private on ruby level.  Otherwise it's never possible to decide what's backward-compatible...

(Of course people are welcome to open issues proposing to document something and make it public API)
